src/dataframe/dataframe.jl

```diff
@@ -1659,3 +1667,34 @@ function allcombinations(::Type{DataFrame}, pairs::Pair{Symbol, <:Any}...)
     @assert size(out_df) == (target_rows, length(colnames))
     return out_df
 end
+
+function metadata(df::DataFrame)
+    meta = getfield(df, :metadata)
+    meta === nothing || return meta
+    new_meta = Dict{String, Any}()
+    setfield!(df, :metadata, new_meta)
+    return new_meta
+end
+
+function hasmetadata(df::DataFrame)
+    meta = getfield(df, :metadata)
+    return !(meta === nothing || isempty(meta))
+end
+
+function metadata(df::DataFrame, col::ColumnIndex)
+    idx = index(df)[col]
+    cols_meta = getfield(df, :column_metadata)
+    if cols_meta === nothing
+        meta = Dict{Int, Dict{String, Any}}()
+        setfield!(df, :column_metadata, meta)
+    else
+        meta = cols_meta
+    end
+    return get!(meta, idx, Dict{String, Any}())
```

Comment:
This avoids creating a dict when it already exists:
```suggestion
    return get!(meta, idx) do
        Dict{String, Any}()
    end
```